File: ruby/easy/palindrome_linked_list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {Boolean}
def is_palindrome(head)
    return true unless head
    center,runner = head,head
    while(runner && runner.next)
        center = center.next
        runner = runner.next.next
    end
    current,runner = center,center.next
    center.next = nil
    while(runner)
        temp = runner.next
        runner.next = current
        current = runner
        runner = temp
    end
    while(head && current)
        return false if(head.val != current.val)
        head = head.next
        current = current.next
    end
    true
end

File: javascript/easy/palindromeLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    return checkPalindrome(head, head, { node: undefined });
};

var checkPalindrome = function(current, runner, obj) {
    if(runner === null || runner.next === null) {
        obj.node = (runner === null) ? current : current.next;
        return true;
    }
    let isPalindrome = checkPalindrome(current.next, runner.next.next, obj) && (current.val === obj.node.val);
    obj.node = obj.node.next;
    return isPalindrome;
}